File: 0335-Self Crossing/0335-Self Crossing.cpp
class Solution {
public:
    bool isSelfCrossing(vector<int>& x) {
        int m = x.size();
        if (m <= 3)
        {
            return false;
        }
        
        for (int i = 3; i < m; ++i)
        {
            // Line 4 crosses line 1.
            if (x[i] >= x[i - 2] && x[i - 3] >= x[i - 1])
            {
                return true;
            }
            
            // Line 5 crosses line 1.
            if (i > 3)
            {
                if (x[i - 1] == x[i - 3] && x[i] + x[i - 4] >= x[i - 2])
                {
                    return true;
                }
            }
            
            // Line 6 crosses line 1.
            if (i > 4)
            {
                if (x[i - 2] >= x[i - 4] && x[i] + x[i - 4] >= x[i - 2] && x[i - 3] >= x[i - 1] && x[i - 5] + x[i - 1] >= x[i - 3])
                {
                    return true;
                }
            }
        }
        
        return false;
    }
};

File: 0301-0400/0335-Self Crossing/0335-Self Crossing.py
class Solution:
    def isSelfCrossing(self, x: List[int]) -> bool:
        if len(x) < 4:
            return False
        for i in range(3, len(x)):
            # Line 4 crosses line 1.
            if x[i] >= x[i - 2] and x[i - 3] >= x[i - 1]:
                return True
            
            # Line 5 crosses line 1.
            if i > 3 and x[i - 1] == x[i - 3] and x[i] + x[i - 4] >= x[i - 2]:
                return True
            
            # Line 6 crosses line 1.
            if (i > 4 and x[i - 2] > x[i - 4] and x[i] + x[i - 4] >= x[i - 2] and
                x[i - 3] > x[i - 1] and x[i - 5] + x[i - 1] >= x[i - 3]):
                return True
        return False
